modules/CbChargeSOMDriver/chargesom/CbChargeSOM.cpp

```diff
@@ -0,0 +1,260 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright chargebyte GmbH and Contributors to EVerest
+#include <cerrno>
+#include <chrono>
+#include <cmath>
+#include <cstring>
+#include <iostream>
+#include <string>
+#include <system_error>
+#include <thread>
+#include <generated/types/cb_board_support.hpp>
+#include "CbChargeSOM.hpp"
+#include "uart.h"
+#include "cb_protocol.h"
+
+using namespace std::chrono_literals;
+
+CbChargeSOM::CbChargeSOM(void) {
+    // clear the structs before usage
+    memset(&this->sc, 0, sizeof(this->sc));
+    memset(&this->uart, 0, sizeof(this->uart));
+
+    // set to invalid fd so that destructor knows whether it must close something
+    this->uart.fd = -1;
+}
+
+CbChargeSOM::~CbChargeSOM() {
+    if (this->uart.fd != -1) {
+        uart_close(&this->uart);
+    }
+}
+
+void CbChargeSOM::init(const std::string& serial_port, bool is_pluggable) {
+    int rv;
+
+    // remember this setting
+    this->is_pluggable = is_pluggable;
+
+    // open the configured device with well-known baudrate
+    rv = uart_open(&this->uart, serial_port.c_str(), 115200);
+    if (rv) {
+        throw std::system_error(errno, std::generic_category(), "Failed to open '" + serial_port + "'");
+    }
+}
+
+void CbChargeSOM::cp_get_values(int& positive_value, int& negative_value) {
+    // this method must be called with `sc_mutex`already hold
+    positive_value = 1000 * (3.3 * this->sc.cp_pos_peak_det * (47.0 + 150.0)) / (4096 * 47.0);
```

Comment:
We should try to get the rid of these magic numbers.

Reply:
This willl go away later since we should received milli-voltages from the safety controller. So I'd stay with it for now if you agree.